ckan/config/middleware/flask_app.py

```diff
@@ -57,17 +231,61 @@ def can_handle_request(self, environ):
         Decides whether it can handle a request with the Flask app by
         matching the request environ against the route mapper
 
-        Returns (True, 'flask_app') if this is the case.
-        '''
+        Returns (True, 'flask_app', origin) if this is the case.
 
-        # TODO: identify matching urls as core or extension. This will depend
-        # on how we setup routing in Flask
+        `origin` can be either 'core' or 'extension' depending on where
+        the route was defined.
+        '''
 
         urls = self.url_map.bind_to_environ(environ)
         try:
-            endpoint, args = urls.match()
-            log.debug('Flask route match, endpoint: {0}, args: {1}'.format(
-                endpoint, args))
-            return (True, self.app_name)
+            rule, args = urls.match(return_rule=True)
+            origin = 'core'
+            if hasattr(rule, 'ckan_core') and not rule.ckan_core:
+                origin = 'extension'
+            log.debug('Flask route match, endpoint: {0}, args: {1}, '
+                      'origin: {2}'.format(rule.endpoint, args, origin))
+            return (True, self.app_name, origin)
         except HTTPException:
             return (False, self.app_name)
+
+    def register_extension_blueprint(self, blueprint, **kwargs):
+        '''
+        This method should be used to register blueprints that come from
+        extensions, so there's an opportunity to add extension-specific
+        options.
+
+        Sets the rule property `ckan_core` to False, to indicate that the rule
+        applies to an extension route.
+        '''
+        self.register_blueprint(blueprint, **kwargs)
+
+        # Get the new blueprint rules
+        bp_rules = [v for k, v in self.url_map._rules_by_endpoint.items()
+                    if k.startswith(blueprint.name)]
+        bp_rules = list(itertools.chain.from_iterable(bp_rules))
+
+        # This compare key will ensure the rule will be near the top.
+        top_compare_key = False, -100, [(-2, 0)]
+        for r in bp_rules:
+            r.ckan_core = False
+            r.match_compare_key = lambda: top_compare_key
+
+
+def _register_core_blueprints(app):
+    u'''Register all blueprints defined in the `views` folder
+    '''
+    views_path = os.path.join(os.path.dirname(__file__),
+                              u'..', u'..', u'views')
+    module_names = [f.rstrip(u'.py')
+                    for f in os.listdir(views_path)
+                    if f.endswith(u'.py') and not f.startswith(u'_')]
+    blueprints = []
+    for name in module_names:
+        module = importlib.import_module(u'ckan.views.{0}'.format(name))
+        blueprints.extend([m for m in inspect.getmembers(module)
+                           if isinstance(m[1], Blueprint)])
+    if blueprints:
```

Comment:
No need to check for a non-empty list here, `for blueprint in blueprints` on an empty list is a nop anyways.